### ISSUE TITLE
Calling `hide()` on $q.dialog with custom component fails

### ISSUE THREAD

**Issue by accron-1** [quasarframework/quasar#10821]
**Describe the bug**
Calling `hide()` on the built-in `$q.dialog` works as expected, but the same call on dialog that has a custom component fails with the error:

```javascript
(index):464 TypeError: dialogRef.value.hide is not a function
    at Proxy.hide (global-dialog.js?0bab:73)
    at Object.eval [as callback] (Todos.vue?c70e:630)
    at Object.cap.fromNative ((index):436)
    at <anonymous>:1:18
```

-----

Based on the following code:
https://quasar.dev/quasar-plugins/dialog#composition-api-variant
I have created the following custom component for a `$q.dialog`:

```javascript
//*** AddItemDialog.vue ***
<template>
  <q-dialog ref="dialogRef" @hide="onDialogHide">
    <q-card class="q-dialog-plugin">
      <q-card-section>
        <div class="text-h6">Add item</div>
      </q-card-section>
      <q-card-section class="q-py-none">
        <div>Type item name</div>
      </q-card-section>

      <q-card-section class="q-pt-none">
        <q-input ref="input" :color="color" dense v-model="todo" autofocus></q-input>
      </q-card-section>

      <q-card-actions align="right">
        <q-btn flat label="Cancel" :color="color" @click="onCancel" v-close-popup></q-btn>
        <q-btn flat label="Add another" :color="color" @click="onAddAnother"></q-btn>
        <q-btn flat label="Add" :color="color" @click="onAdd" v-close-popup></q-btn>
      </q-card-actions>
    </q-card>
  </q-dialog>
</template>

<script setup>
  import { useDialogPluginComponent, useQuasar } from 'quasar';
  import { ref, computed } from 'vue';

  const props = defineProps({
    callback: {
      type: Function,
      required: true
    }
  });
  const emits = defineEmits( [
    // REQUIRED; need to specify some events that your
    // component will emit through useDialogPluginComponent()
    ...useDialogPluginComponent.emits,
  ]);
  const todo = ref(''),
    input = ref(null),
    $q = useQuasar();
  const { dialogRef, onDialogHide, onDialogOK, onDialogCancel } = useDialogPluginComponent();

  function onCancel() {
    todo.value = '';
    props.callback('cancel', todo.value);
  }

  function onAddAnother() {
    input.value.focus();
    props.callback('addanother', todo.value);
    todo.value = '';
  }

  function onAdd() {
    props.callback('add', todo.value);
  }

  function onOKClick () {
    // on OK, it is REQUIRED to
    // call onDialogOK (with optional payload)
    // console.log('onok');
    onDialogOK()
    // or with payload: onDialogOK({ ... })
    // ...and it will also hide the dialog automatically
  }

  // we can passthrough onDialogCancel directly
  const onCancelClick = onDialogCancel;
  
  const color = computed(() =>{
    return $q.dark.isActive ? 'warning' : 'primary';
  });
</script>
```

I then use the above component in a `$q.dialog` as follows (based on this: https://quasar.dev/quasar-plugins/dialog#invoking-custom-component):
```javascript
const activeDialog = ref(null);
...
activeDialog.value = $q.dialog({
        component: AddItemDialog,
        componentProps: {
          callback: (btnClicked, input) => {
            if (btnClicked === 'cancel') {
              activeDialog.value = null;
              return;
            }
            //do something...
          }
        },
      });
```

When `activeDialog.value.hide()` is called, while the dialog is open, the following error happens:
```javascript
(index):464 TypeError: dialogRef.value.hide is not a function
    at Proxy.hide (global-dialog.js?0bab:73)
    at Object.eval [as callback] (Todos.vue?c70e:630)
    at Object.cap.fromNative ((index):436)
    at <anonymous>:1:18
```

`console.log(activeDialog.value.hide)` gives:
```javascript
hide() {
        if (dialogRef.value !== null) {
          dialogRef.value.hide();
        }

        return API;
      }
```

So there does seem to be a `hide()` function.

**Codepen/jsFiddle/Codesandbox (required)**
It's Capacitor project, I don't know how to wire the equivalent in Codepen

**To Reproduce**
See above

**Expected behavior**
Calling `hide()` on a `$q.dialog` with custom component should behave the same as for the default dialog.

**Screenshots**
NA

**Platform (please complete the following information):**
Quasar Version:
@quasar/app Version: 3

Quasar mode:
  - [ ] SPA
  - [ ] SSR
  - [ ] PWA
  - [ ] Electron
  - [ ] Cordova
  - [ X] Capacitor
  - [ ] BEX

Tested on:
  - [ ] SPA
  - [ ] SSR
  - [ ] PWA
  - [ ] Electron
  - [ ] Cordova
  - [ X] Capacitor
  - [ ] BEX

OS: Ubuntu 21.04
Node: 12.22.1
NPM: 12.22.1
Android: SDK 29

**Additional context**
I use the `<script setup>` style for all my Vue components

**Comment by metalsadman** [comment]
try awaiting a nextTick before invoking your dialog ref call.

**Comment by accron-1** [comment]
Thank you for the suggestion. Unfortunately it doesn't fix the issue.

The use case is that I handle the back button event: if the dialog is not open then I go back to the previous screen on the first click.
If the dialog *is* open then the first click should close the dialog (=`activeDialog.value.hide()`), and the second click returns to the previous screen.

I probably should have explained the use case in the original post...

---

One thing I noticed in the docs is that the Options API variant of the custom component (https://quasar.dev/quasar-plugins/dialog#options-api-variant) has a `hide()` method, that does `this.$refs.dialog.hide()`.
As far as I see/understand there is no equivalent code in the Composition API example of the code, but maybe it's supposed to work even so?

**Comment by metalsadman** [comment]
I've tested it and it works as expected in codepen here https://codepen.io/metalsadman/pen/LYLqyrQ?editors=1010 and also in my actual project as long as you use the provided composables, although this also works if you do it manually same as you see in the options api example. will try to make a stackblitz with `script setup` as i suspect this is what's at fault here.

**Comment by metalsadman** [comment]
can confirm that it doesn't work in `script setup`, global dialog needs adjusting.

**Comment by accron-1** [comment]
Thank you for looking into it, and patching it so quickly!

I will use `setup()` in the meantime. When do you think the patch will be officially released?

**Comment by rstoenescu** [comment]
Will be available in Quasar v2.1.1
